pkg/volume/cinder/cinder.go

```diff
@@ -0,0 +1,281 @@
+	AttachDisk(builder *cinderVolumeBuilder, globalPDPath string) error
+	// Detaches the disk from the kubelet's host machine.
+	DetachDisk(cleaner *cinderVolumeCleaner) error
+}
+
+var _ volume.Builder = &cinderVolumeBuilder{}
+
+type cinderVolumeBuilder struct {
+	*cinderVolume
+	fsType             string
+	readOnly           bool
+	blockDeviceMounter mount.Interface
+}
+
+// cinderPersistentDisk volumes are disk resources provided by C3
+// that are attached to the kubelet's host machine and exposed to the pod.
+type cinderVolume struct {
+	volName string
+	podUID  types.UID
+	// Unique identifier of the volume, used to find the disk resource in the provider.
+	pdName string
+	// Filesystem type, optional.
+	fsType string
+	// Specifies the partition to mount
+	//partition string
+	// Specifies whether the disk will be attached as read-only.
+	readOnly bool
+	// Utility interface that provides API calls to the provider to attach/detach disks.
+	manager cdManager
+	// Mounter interface that provides system calls to mount the global path to the pod local path.
+	mounter mount.Interface
+	// diskMounter provides the interface that is used to mount the actual block device.
+	blockDeviceMounter mount.Interface
+	plugin             *cinderPlugin
+}
+
+func detachDiskLogError(cd *cinderVolume) {
+	err := cd.manager.DetachDisk(&cinderVolumeCleaner{cd})
+	if err != nil {
+		glog.Warningf("Failed to detach disk: %v (%v)", cd, err)
+	}
+}
+
+func (b *cinderVolumeBuilder) SetUp() error {
+	return b.SetUpAt(b.GetPath())
+}
+
+// SetUp attaches the disk and bind mounts to the volume path.
+func (b *cinderVolumeBuilder) SetUpAt(dir string) error {
+	// TODO: handle failed mounts here.
+	notmnt, err := b.mounter.IsLikelyNotMountPoint(dir)
+	glog.V(4).Infof("PersistentDisk set up: %s %v %v", dir, !notmnt, err)
+	if err != nil && !os.IsNotExist(err) {
+		return err
+	}
+	if !notmnt {
+		return nil
+	}
+	globalPDPath := makeGlobalPDName(b.plugin.host, b.pdName)
+	if err := b.manager.AttachDisk(b, globalPDPath); err != nil {
+		return err
+	}
+
+	options := []string{"bind"}
+	if b.readOnly {
+		options = append(options, "ro")
+	}
+
+	if err := os.MkdirAll(dir, 0750); err != nil {
+		// TODO: we should really eject the attach/detach out into its own control loop.
+		detachDiskLogError(b.cinderVolume)
+		return err
+	}
+
+	// Perform a bind mount to the full path to allow duplicate mounts of the same PD.
+	err = b.mounter.Mount(globalPDPath, dir, "", options)
+	if err != nil {
+		notmnt, mntErr := b.mounter.IsLikelyNotMountPoint(dir)
+		if mntErr != nil {
+			glog.Errorf("IsLikelyNotMountPoint check failed: %v", mntErr)
+			return err
+		}
+		if !notmnt {
+			if mntErr = b.mounter.Unmount(dir); mntErr != nil {
+				glog.Errorf("Failed to unmount: %v", mntErr)
+				return err
+			}
+			notmnt, mntErr := b.mounter.IsLikelyNotMountPoint(dir)
+			if mntErr != nil {
+				glog.Errorf("IsLikelyNotMountPoint check failed: %v", mntErr)
```

Comment:
Again, the directory this is relevant to should be included in the error message.